### PR TITLE
CDAP-3263 Removing row.case.insensitive field and adding a field case…

### PR DESCRIPTION
**Pull request by gokulavasan**
… choice in DBSource plugin

JIRA : https://issues.cask.co/browse/CDAP-3263

Build : http://builds.cask.co/browse/CDAP-RBT458

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchETLDBTest.java [hunk 1]
```diff
@@ -300,12 +307,24 @@ public void testDBSourceWithCaseSensitiveRowKey() throws Exception {
     MapReduceManager mrManager = appManager.getMapReduceManager(ETLMapReduce.NAME);
     mrManager.start();
     mrManager.waitForFinish(5, TimeUnit.MINUTES);
+    List<RunRecord> runRecords = mrManager.getHistory();
+    Assert.assertNotEquals(ProgramRunStatus.FAILED, runRecords.get(0).getStatus());
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchETLDBTest.java]: can we assert succeeded here?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchETLDBTest.java [hunk 2]
```diff
@@ -339,6 +358,8 @@ public void testDBSink() throws Exception {
...
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchETLDBTest.java]: ditto

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java [hunk 3]
```diff
@@ -60,6 +64,11 @@
     "import from the specified table. Examples: SELECT COUNT(*) from <my_table> where <my_column> 1, " +
     "SELECT COUNT(my_column) from my_table. NOTE: Please include the same WHERE clauses in this query as the ones " +
     "used in the import query to reflect an accurate number of records to import.";
+  private static final String COLUMN_CASE_DESCRIPTION = "Sets the case of the column names returned from the query." +
+    "Possible options are UPPER, lower. By default or for any other input, the column names are not modified" +
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java]: I can see why you did this, but is it confusing in a description to have "UPPER, lower"? Does that give the notion that those are the only two supported values (whereas the code accepts any case for these two)

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+   * @return {@link StructuredRecord} which contains field names confirming to the {@link FieldCase} passed in
+   * @throws Exception if there is a conflict in the field names while converting the case
+   */
+  public static StructuredRecord convertCase(StructuredRecord input, FieldCase fieldCase) throws Exception {
+    if (fieldCase.equals(FieldCase.NONE)) {
+      return input;
+    }
+
+    Schema oldSchema = input.getSchema();
+    Map<String, String> fieldNameMap = Maps.newHashMap();
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: ``new HashMap<>();``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    for (Map.Entry<String, String> nameMap : fieldNameMap.entrySet()) {
+      recordBuilder.set(nameMap.getValue(), input.get(nameMap.getKey()));
+    }
+    return recordBuilder.build();
+  }
+
+  private StructuredRecordUtils() {
+  }
+
+  private static String nameChange(String oldName, FieldCase fieldCase) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: ``changeName()``, to make it a verb?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/common/StructuredRecordUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  @Test
+  public void testLowerAndUpperCaseTransformation() throws Exception {
+    StructuredRecord record = StructuredRecord.builder(
+      Schema.recordOf("dbrecord",
+                      Schema.Field.of("Name", Schema.nullableOf(Schema.of(Schema.Type.STRING))),
+                      Schema.Field.of("Age", Schema.of(Schema.Type.INT)))).set("Name", "Abcd").set("Age", 10).build();
+    StructuredRecord upperCaseRecord = StructuredRecordUtils.convertCase(
+      record, FieldCase.toFieldCase("upPer"));
+    Assert.assertEquals("Abcd", upperCaseRecord.get("NAME"));
+    Assert.assertEquals(10, upperCaseRecord.get("AGE"));
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/common/StructuredRecordUtilsTest.java]: should we assert that the structured record does not have the old fields?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/common/StructuredRecordUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    StructuredRecord lowerCaseRecord = StructuredRecordUtils.convertCase(
+      record, FieldCase.toFieldCase("lowEr"));
+    Assert.assertEquals("Abcd", lowerCaseRecord.get("name"));
+    Assert.assertEquals(10, lowerCaseRecord.get("age"));
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/common/StructuredRecordUtilsTest.java]: ditto

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    Schema oldSchema = input.getSchema();
+    Map<String, String> fieldNameMap = Maps.newHashMap();
+    List<Schema.Field> newFields = Lists.newArrayList();
+    for (Schema.Field field : oldSchema.getFields()) {
+      String newName = nameChange(field.getName(), fieldCase);
+      if (fieldNameMap.containsValue(newName)) {
+        // field name used already. indication of field names conflict. can't do anything.
+        throw new Exception(String.format(
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: Do we have an ``InvalidConfigurationException`` or something like that? If the exception is checked (expected to be caught by callers), then better to throw a specific exception? If we don't care about clients catching it, then should we just throw ``IllegalStateException``?

**Comment by bdmogal** [comment]
@gokulavasan only minor comments.

File: cdap-ui/templates/common/Database.json [hunk 9]
```diff
@@ -59,7 +59,15 @@
           "countQuery" : {
              "widget": "textbox",
              "label": "Count SQL"
-          }
+          },
+	  "columnNameCase" : {
+	     "widget": "select",
+	     "label" : "Column Case",
+	     "properties" : {
+		"values" : ["UPPER", "lower", "No change"],
```
Comment (bdmogal) [inline cdap-ui/templates/common/Database.json]: nit: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java (hunk 3 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java]: Oh hmm. the code accepts any case of upper, lower. Should we just check for one specific case?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java (hunk 4 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: Are we moving away from Guava classes? Or just want to reduce dependencies in new code?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java (hunk 8 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: I looked for a good type of exception to throw. There is ``InvalidPluginConfigurationException`` but that is only in cdap-app-fabric. May we can move it elsewhere.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: yeah, with java 7, the ``new HashMap<>()`` is concise, so we're moving away

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java]: no, I just meant in the description does it throw you off (that these are the only two values supported). I saw that the code accepts any case, which is good.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/common/StructuredRecordUtils.java]: yeah, not sure if that will increase the size of this PR, but that sounds like a good exception. Is that exception supposed to be used elsewhere?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java (hunk 3 above)
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java]: These are the two options that users will see on the UI tool tip. Can you suggest an alternate description?

Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/source/DBSource.java]: upper, lower?

(Only if you think the current description could throw users off). I don't feel too strongly about it.

**Comment by gokulavasan** [comment]
@bdmogal Addressed PR comments. Please take a look again.

**Comment by bdmogal** [comment]
LGTM

**Comment by gokulavasan** [comment]
Build is :green_apple:. Merging. Thanks for the review @bdmogal.

**Pull request merged**